Add persistent sessions

AngellusMortis opened the pull request:
Something I have wanted to support for a file since UniFi OS gets funky and returns 401 as a rate limit sometimes. 

* Adds `platformdirs` for getting the correct user cache / config directories
* Stores and loads `TOKEN` cookies to disk so `ProtectApiClient` can have persistent sessions. If you run `unifi-protect` 5-10 times rapidly, it will not randomly get a 401 since it does not issue a login each time. 
* Refactors how headers are handed so `cookie` / `x-csrf-token` are not randomly accidently deleted
* Fixes deprecation warning in CLI interface (replace `asyncio.get_event_loop` with `asyncio.run` for newer versions of Python)
* Set `x-csrf-token` on every request

This will require a HA side change to redirect the config directory to correct location.

AngellusMortis commented on pyunifiprotect/api.py:
```diff
@@ -174,6 +190,10 @@ def __init__(
         self._verify_ssl = verify_ssl
         self._ws_timeout = ws_timeout
 
+        self.config_dir = config_dir or (Path(user_config_dir()) / "ufp")
+        self.cache_dir = cache_dir or (Path(user_cache_dir()) / "ufp_cache")
```
It does not seem like these do disk IO: https://github.com/platformdirs/platformdirs/blob/main/src/platformdirs/unix.py

bdraco replied:
`expanduser` calls `pwd.getpwuid` and `pwd.getpwnam` both of which do blocking I/O to read /etc/passwd, etc

AngellusMortis replied:
`/etc/passwd` is likely going to be heavily cached by the OS those because it is an incredibly high access file, so I do not really expect that to be an issue. It still has to take the hit somewhere though. Either when ProtectApiClient is initialized or when it is imported, both are likely to be inside of an event loop.

bdraco commented on pyunifiprotect/api.py:
```diff
@@ -439,26 +480,88 @@ async def authenticate(self) -> None:
+    async def _read_auth_config(self) -> SimpleCookie | None:
+        """Read auth cookie from config."""
+
+        if not await aos.path.isfile(self.config_file):
```
Path.isfile does a blocking stat

AngellusMortis replied:
That is why it is using aiofile instead of os.

bdraco replied:
Sorry I missed it was wrapped in aiofiles

bdraco commented on pyunifiprotect/api.py:
```diff
@@ -439,26 +480,88 @@ async def authenticate(self) -> None:
+        for key, value in session.get("metadata", {}).items():
+            cookie["TOKEN"][key] = value
+
+        cookie_value = str(cookie["TOKEN"]).replace("Set-Cookie: ", "")
```
This would break if `Set-Cookie: ` changes case

AngellusMortis replied:
It is core functionality of `SimpleCookie` so it is not likely to change. I cannot use `lower` here because that would break the JWT. I could change it to a case insensitive regex match, but that is slower.

bdraco replied:
I think its fine. Only pointing it out as due-diligence

AngellusMortis replied:
Thinking more about it, I think the regex might be a bit faster since it does not need to scan the whole string (`^` modifier).

bdraco commented on pyunifiprotect/api.py:
```diff
@@ -439,26 +480,88 @@ async def authenticate(self) -> None:
+    async def _read_auth_config(self) -> SimpleCookie | None:
+        """Read auth cookie from config."""
+
+        if not await aos.path.isfile(self.config_file):
+            _LOGGER.debug("no config file, not loading session")
+            return None
+
+        async with aiofiles.open(self.config_file, "rb") as f:
```
Not that it likely matters here, but the common case here is that the file will exist so skipping the stat and trying to open the file (and handling the exception if it doesn't) will save one executor job which would reduce the latency here

bdraco commented:
Testing this one prod now

bdraco commented on pyunifiprotect/api.py:
```diff
@@ -597,7 +699,6 @@ def __init__(
         ignore_stats: bool = False,
         ignore_unadopted: bool = True,
         debug: bool = False,
-        cache_dir: Optional[Path] = None,
```
I guess we still need to accept a `cache_dir` and `config_dir` here so we can pass it to the super so HA can override the location

AngellusMortis replied:
It was moved to the base class.

bdraco replied:
Is the plan to no longer use this class in HA?

```
2024-03-21 08:19:31.109 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Maui UNVR for unifiprotect
Traceback (most recent call last):                                                                                                                                                                                                                        
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 549, in async_setup
    result = await component.async_setup_entry(hass, self)                                                                                                                                                                                                
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                
  File "/usr/src/homeassistant/homeassistant/components/unifiprotect/__init__.py", line 67, in async_setup_entry
    protect = async_create_api_client(hass, entry)                                                                                                                                                                                                        
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                        
  File "/usr/src/homeassistant/homeassistant/components/unifiprotect/utils.py", line 136, in async_create_api_client
    return ProtectApiClient(                                                                                                                                                                                                                              
           ^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                              
TypeError: ProtectApiClient.__init__() got an unexpected keyword argument 'cache_dir'  
```

AngellusMortis replied:
Whoops. I guess there is not a kwargs catch on there.

AngellusMortis commented:
Added the kind of talked about suggestions and also added csrf tokens so they get stored as well. Otherwise, if you do a post right away (like via the CLI), it could cause a 403.

bdraco commented:
Pulling and re-testing now
